File: 03-sliding_window/00424-Longest_Repeating_Character_Replacement/00424-longest_repeating_character_replacement.go
// 424: Longest Repeating Character Replacement
// https://leetcode.com/problems/longest-repeating-character-replacement/

package main

import (
	"fmt"
	"math"
)

// SOLUTION
func characterReplacement(s string, k int) int  {
	count := make([]int, 26)
	maxCount := 0
	i, j := 0, 0

	result := 0

	for j<len(s) {
		count[s[j]-'A']++;
		maxCount = int(math.Max(float64(maxCount), float64(count[s[j]-'A'])))

		if j-i+1-maxCount > k {
			count[s[i]-'A']--
			i++
		}
		result = int(math.Max(float64(result), float64(j-i+1)))
		j++
	}

	return result
}


func main() {
	// INPUT
	s := "ABAB"
	k := 2

	// OUTPUT
	result := characterReplacement(s, k)
	fmt.Println(result)
}

File: 03-sliding_window/00424-Longest_Repeating_Character_Replacement/00424-longest_repeating_character_replacement.java
// 424: Longest Repeating Character Replacement
// https://leetcode.com/problems/longest-repeating-character-replacement/


class Solution {
    // SOLUTION
    int characterReplacement(String s, int k) {
        int[] count = new int[26];
        int maxCount = 0;
        int i = 0;
        int j = 0;

        int result = 0;

        while (j<s.length()) {
            count[s.charAt(j)-'A']++;
            maxCount = Math.max(maxCount, count[s.charAt(j)-'A']);

            if (j-i+1-maxCount > k) {
                count[s.charAt(i)-'A']--;
                i++;
            }

            result = Math.max(result, j-i+1);
            j++;
        }

        return result;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        String s = "ABAB";
        int k = 2;

        // OUTPUT
        var result = o.characterReplacement(s, k);
        System.out.println(result);
    }
}